python/tk_multi_loader/dialog.py

```diff
@@ -1265,6 +1280,94 @@ def _load_entity_presets(self):
         # data has properly arrived in the model.
         self._on_home_clicked()
 
+    def _setup_hierarchy_model(self, app, root):
+        """
+        Create the model and proxy model required by a hierarchy type configuration setting.
+
+        :param app: :class:`Application`, :class:`Engine` or :class:`Framework` bundle instance
+                    associated with the loader.
+        :param root: The path to the root of the Shotgun hierarchy to display.
+        :return: Created `(model, proxy model)`.
+        """
+
+        # Resolve any magic tokens in the root.
+        if app.context.project:
+            project_id = app.context.project.get("id")
+            root = root.replace("{context.project.id}", str(project_id))
+        else:
+            root = None
+
+        # Construct the hierarchy model.
+        model = SgHierarchyModel(self, bg_task_manager=self._task_manager)
+
+        # Create a proxy model.
+        proxy_model = QtGui.QSortFilterProxyModel(self)
+        proxy_model.setSourceModel(model)
+
+        # Impose and keep the sorting order on the default display role text.
+        proxy_model.sort(0)
+        proxy_model.setDynamicSortFilter(True)
+
+        # We need to provide a dictionary that identifies what additional fields to include
+        # for the loaded hierarchy leaf entities in addition to "id" and "type".
+        # When available, these fields will be used to add info in the detail panel.
+        # TODO: Our entity type list for entities with publishes should be retrieved from the settings.
```

Comment:
The issue is not to decide what fields are available for each entity, but what entities to put as keys in the `entity_fields` dictionary.

When a requested field is not available for a specific entity, it is simply ignored by the Shotgun model. This is already how the query model is working with such fields:
https://github.com/shotgunsoftware/tk-multi-loader2/blob/master/python/tk_multi_loader/model_entity.py#L54

The Loader already gracefully handles the lack of some fields when displaying the detail panel.